### DISCUSSION TITLE
Bad image from consecutive histogram equalization AdaptiveTileInterpolation

### DISCUSSION THREAD

**Discussion by leonyu1010** [SixLabors/ImageSharp#1640] [Q&A]
Hi
I use the follow code to test  consecutive histogram Equalizatio + resize.  I expect that each cycle is independent and I get near-identical result each time. However, the attached screenshot shows the first cycle is fine, and it degrades in subsequent cycles

Is this a correct way of using the library ? and 
Is there a way to "reset"  after each cycle so that I can always get the result as good as the first cycle ?

```c#         
// run the same image and same pipeline multiple times

for (int cycle = 0; cycle < 5; cycle++)
{
    //  pixels  is a SixLabors.ImageSharp.PixelFormats.L16  image   width = 5120, height = 9234
    unsafe
    {
        fixed (void* pixPtr = pixels)
        {
           // the wrappedImage intance is verfied to be correct
            using var wrappedImage = SixLabors.ImageSharp.Image.WrapMemory<SixLabors.ImageSharp.PixelFormats.L16>(pixPtr, image.Width, image.Height);

            var options = new HistogramEqualizationOptions
            {
                Method = HistogramEqualizationMethod.AdaptiveTileInterpolation,
                LuminanceLevels = 4096,
                ClipHistogram = false,
                ClipLimit = 350,
                NumberOfTiles = 8
            };

            // the processing pipeline:  histogram equalization  and reduce image size. 
            var processed = wrappedImage.Clone(ctx =>
            {
                ctx.HistogramEqualization(options);
                ctx.Resize(image.Width / 4, image.Height / 4, KnownResamplers.Bicubic);
            });

           //  save the processed  image to JPEG
            processed.SaveAsJpeg(filename, new JpegEncoder() {Quality = 80, Subsample = JpegSubsample.Ratio444});
        }
    }
}
```

in the following screenshot,
from left to right:  original image,  1st cycle result, 2nd cycle result ...  and 5th cycle result

![Screenshot from 2021-05-28 16-23-59](https://user-images.githubusercontent.com/11080856/119955922-208c4380-bfd3-11eb-9740-10e924966438.png)

**Comment by JimBobSquarePants** [comment]
Hi @leonyu1010 where does the `pixels` buffer come from? Is it immutable?

**Reply by leonyu1010** [reply]
for this test program, pixels is loaded a TIFF image from disk.  the buffer is not accessed elsewhere. 
```c#
for (cycle = 0; cycle < 5; cycle++)
{
  // open the image file from disk
  // call a TIFF decoder
  // result is placed to pixels
   pixels =  load TIFF from disk.  so pixels is a buffer of  uint16
   
  // then  wrap pixel as PixelFormats.L16  image, namely wrappedImage in the code snippet .
  //  here  i saved wrappedImage to jpeg and it looks correct. 
  // run pipeline
  // save jpeg
}
```

**Reply by antonfirsov** [reply]
Your ImageSharp code looks fine (with the exception of a missing `using` before `var processed`, but it's unlikely to be related) I  tried if I can reproduce this in our test suite but I couldn't, the following code outputs the same image 5 times:

<details>

```C#
        [Fact]
        public unsafe void Repro()
        {
            string path = TestFile.GetInputFileFullPath(TestImages.Jpeg.Baseline.Cmyk);
            using var image = Image.Load<L16>(path);
            string reproDir = TestEnvironment.CreateOutputDirectory("HistogramRepro");

            image.TryGetSinglePixelSpan(out Span<L16> pixels);
            for (int i = 0; i < 5; i++)
            {
                string fileName = Path.Combine(reproDir, $"{i}.jpg");

                fixed (void* pixPtr = pixels)
                {
                    // the wrappedImage intance is verfied to be correct
                    using var wrappedImage = Image.WrapMemory<L16>(pixPtr, image.Width, image.Height);

                    var options = new HistogramEqualizationOptions
                    {
                        Method = HistogramEqualizationMethod.AdaptiveTileInterpolation,
                        LuminanceLevels = 4096,
                        ClipHistogram = false,
                        ClipLimit = 350,
                        NumberOfTiles = 8
                    };

                    Image<L16> processed = wrappedImage.Clone(ctx =>
                    {
                        ctx.HistogramEqualization(options);
                        ctx.Resize(image.Width / 4, image.Height / 4, KnownResamplers.Bicubic);
                    });

                    processed.SaveAsJpeg(fileName, new JpegEncoder() { Quality = 80, Subsample = JpegSubsample.Ratio444 });
                }
            }
        }
```

</details>

@leonyu1010 any chance you can create a standalone minimal repro for this? ~Otherwise my only explanation is that the content in your `pixels` is changing independently of the processor code.~ *Edit: just seen your reply, a minimal self-contained repro would be still valuable.*

**Reply by antonfirsov** [reply]
As a sidenote, we implemented TIFF in #1553 and #1634. With our MyGet nightlies you should be able to load it as any regular image without the need of an external codec. A NuGet release will follow in 1 month timeline hopefully.

**Reply by antonfirsov** [reply]
> pixels =  load TIFF from disk.  so pixels is a buffer of  uint16

Could it be that you are loading different layers of the same image between iterations or something like that?
Can you try saving it without processing?

**Reply by leonyu1010** [reply]
i will try to create a demo repo for this. 

meanwhile this screen shows that the `wrappedImage` before pipeline is correct for each cycle. 
left one is the original 12-bit TIFF.  the other 5  are `wrappedImage.SaveAsJpeg`.  I thought it makes sense it goes darker under 8 bit.  And that is actually the purpose of using histogram equalization here to increase contrast. 

![Screenshot from 2021-05-28 18-33-47](https://user-images.githubusercontent.com/11080856/119971366-74068d80-bfe3-11eb-9783-63681def9ef4.png)

**Reply by JimBobSquarePants** [reply]
@antonfirsov I think the input size is relevant here. I've managed to recreate it using an input png of the same dimensions. 

Things I've discovered so far:

- Using a pinned buffer is irrelevant. Any image of the input size seems to be sufficient to trigger the issue.
- The issue is specific to histogram equalization.
- `HistogramEqualizationMethod.AdaptiveTileInterpolation` is the only one affected.

As a side note. `HistogramEqualizationMethod.AdaptiveSlidingWindow` is unusably slow for larger images taking over 5 minutes to complete on my SB2.

**Reply by leonyu1010** [reply]
thanks for the tips. I get very good result using `HistogramEqualizationMethod.Global`.

**Reply by JimBobSquarePants** [reply]
PR with fix here #1641
